pyop2/compilation.py

```diff
@@ -174,18 +186,54 @@ class Compiler(object):
     :kwarg comm: Optional communicator to compile the code on
         (defaults to COMM_WORLD).
     """
-    def __init__(self, cc, ld=None, cppargs=[], ldargs=[],
-                 cpp=False, comm=None):
-        ccenv = 'CXX' if cpp else 'CC'
+    def __init__(self, extra_cppflags=None, extra_ldflags=None, cpp=False, comm=None):
+        self._extra_cppflags = extra_cppflags or []
+        self._extra_ldflags = extra_ldflags or []
+
+        self._cpp = cpp
+        self._debug = configuration["debug"]
+
         # Ensure that this is an internal communicator.
         comm = dup_comm(comm or COMM_WORLD)
         self.comm = compilation_comm(comm)
-        self._cc = os.environ.get(ccenv, cc)
-        self._ld = os.environ.get('LDSHARED', ld)
-        self._cppargs = cppargs + configuration['cflags'].split()
-        if configuration["use_safe_cflags"]:
-            self._cppargs += self.workaround_cflags
-        self._ldargs = ldargs + configuration['ldflags'].split()
+
+    @property
+    def cc(self):
+        return os.environ.get("PYOP2_CC", self._cxx if self._cpp else self._cc)
+
+    @property
+    def ld(self):
+        return os.environ.get("PYOP2_LD", None)
+
+    @property
+    def cppflags(self):
+        try:
+            return os.environ["PYOP2_CPPFLAGS"]
+        except KeyError:
+            cppflags = self._cppflags + self._extra_cppflags
+
+            if not self._debug:
+                cppflags += self._debugflags
+            else:
+                cppflags += self._optflags
+
+            if self._cpp:
+                cppflags += self._cxxflags
+            else:
+                cppflags += self._cflags
+
+            return cppflags
+
+    @property
+    def ldflags(self):
+        try:
+            return os.environ["PYOP2_LDFLAGS"]
+        except KeyError:
+            return self._ldflags + self._extra_ldflags
```

Comment:
I think I got a bit confused about this last time. Which flags do we consider essential and which are optional? Perhaps `self._ldflags` should be constant (i.e. the flags we will always want to have like `-lm`) and then the environment variable could only overwrite `_extra_ldflags`.